README.md
Tabular
=======

Tabular is a **reader** for data in ascii table format and **test helpers** to facilitate
testing with ascii tables.

Reader
------

The reader allows you to read string data formatted like this:

<pre>
+------------------+--------------------+
| name             | dob                |
+------------------+--------------------+
| Malcolm Reynolds | September 20, 2468 |
| Zoe Washburne    | February 15, 2484  |
+------------------+--------------------+
</pre>

You can add row separators so that column values can wrap over multiple rows. This helps
you manage the width of tables.

<pre>
+-----------+--------------+
| name      | dob          |
+-----------+--------------+
| Malcolm   | September 20,|
| Reynolds  | 2468         |
+-----------+--------------+
| Zoe       | February 15, |
| Washburne | 2484         |
+-----------+--------------+
</pre>        
        
Wrapped cells are folded--the two tables produce the same results.

Test Helpers
------------

The `compare`, `equal?`, and `assert_equal` functions compare two ascii tables. 

**`compare`** generates a matrix from two ascii tables indicating which cells are equal
*and which are different. You can provide an optional list of per column comparators. When
*a comparator is provided for a column it will be used to compare the values of the cells
*for that column. See [test_support_test.exs][2] for examples that include comparators.

[2]: https://github.com/kellyfelkins/tabular/blob/master/test/test_support_test.exs

**`equal?`** takes the matrix generated by compare and returns a boolean indicating if
*original ascii tables are, well, equal.

**`assert_equal`** asserts that the results generated from `compare` indicate the tables
*are equal, and, if not, generates a failure message showing the table with differences
*marked.

In [Ascii Tables For Clearer Testing][1] I discuss using ascii tables to improve
comprehension of software tests.

[1]: https://punctuatedproductivity.wordpress.com/2016/02/02/ascii-tables-for-clearer-testing/

Installation
------------

The package can be installed by adding `tabular` to your list of dependencies in
`mix.exs`:

```elixir
def deps do
  [
    {:tabular, "~> 1.0"}
  ]
end
```

Usage
-----

Rows are returned as either lists of lists or lists of maps.

<pre>
+-----------------------+------------------------------+-----------------------------------+
| **Ascii Table Value** | **Returned Value**           | **Notes**                         |
+-----------------------+------------------------------+-----------------------------------+
| Malcolm Reynolds      | "Malcolm Reynolds"           | Most values returned as string    |
+-----------------------+------------------------------+-----------------------------------+
| 123                   | "123"                        | including numbers                 |
+-----------------------+------------------------------+-----------------------------------+
| wrapped strings are   | "wrapped strings are folded" | Similar to yaml, wrapped          |
| folded                |                              | strings are folded with a single  |
|                       |                              | space replacing the new line      |
+-----------------------+------------------------------+-----------------------------------+
|                       | ""                           | an empty string is returned for   | 
|                       |                              | blank cells                       |
+-----------------------+------------------------------+-----------------------------------+
| nil                   | nil                          | nil, true, and false are          |
|                       |                              | special values                    |
+-----------------------+------------------------------+-----------------------------------+
| true                  | true                         |                                   |
+-----------------------+------------------------------+-----------------------------------+
| false                 | false                        |                                   |
+-----------------------+------------------------------+-----------------------------------+
| :foo                  | :foo                         | Values beginning with a colon are |
|                       |                              | returned as atoms                 |
+-----------------------+------------------------------+-----------------------------------+
</pre>

**Reading a String**

```elixir
  Tabular.to_list_of_lists(table)
  |> Enum.each(fn [col1, col2, col3] = row ->
    # use row or column data here...
  end
```

More examples can be found in the Examples of Testing With ASCII Tables
[repo](https://github.com/kellyfelkins/examples_of_testing_with_ascii_tables).

The docs can be found at [https://hexdocs.pm/tabular](https://hexdocs.pm/tabular).

## License

Blue Oak Model license. Please see [LICENSE][license] for details.

[LICENSE]: https://github.com/kellyfelkins/tabular/blob/master/LICENSE.md


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.1] - 2020-09-15

### Changed

- Improves documentation.

## [1.2.0] - 2020-09-14

### Added

- Adds processing of tables that do not have separator lines.

## [1.1.0] - 2019-07-31

### Added

- This changelog.

### Changed

- Merges [PR](https://github.com/kellyfelkins/tabular/pull/4) from
  [@achoe314](https://github.com/achoe314) that deprecates 
  `_no_header` readers while adding a 
  header: true/false option to readers.
    
## [1.0.0] - 2019-07-30

### Added

- `TestSupport.assert_equal(results_table)` returns true when
  no differences are found in the results table, and does a fail
  with a failure message that displays a table with differences
  highlighted. Thank you [@aarongraham](https://github.com/aarongraham).
  

### Changed

- `TestSupport.compare(table1, table2)` now returns all cell values,
  and tuples in the case where cell contents do not match.
  `TestSupport.equal?()` accepts this new format. Thank you 
  [@aarongraham](https://github.com/aarongraham).



test/tabular_test.exs
defmodule TabularTest do
  use ExUnit.Case, async: true

  doctest Tabular

  describe "has_separators?" do
    test "true when the table has separators" do
      data_as_table = """
            +-----------+--------------------+--------------+
            | name      | dob                | predictable? |
            +-----------+--------------------+--------------+
            | Malcolm   | September 20, 2468 | false        |
            +-----------+--------------------+--------------+
            | Zoe       | February 15, 2484  |              |
            +-----------+--------------------+--------------+
      """

      assert Tabular.has_separators?(data_as_table)
    end

    test "false when the table does not have separators" do
      data_as_table = """
      +-----------+--------------------+--------------+
      | name      | dob                | predictable? |
      +-----------+--------------------+--------------+
      | Malcolm   | September 20, 2468 | false        |
      | Zoe       | February 15, 2484  |              |
      +-----------+--------------------+--------------+
      """

      refute Tabular.has_separators?(data_as_table)
    end
  end

  describe "cell_line_groups" do
    test "when separator lines, extracts lines with cell contents and groups them" do
      data_as_lines = [
        "+-----------+--------------------+",
        "| name      | dob                |",
        "+-----------+--------------------+",
        "| Malcolm   | September 20, 2468 |",
        "| Reynolds  |                    |",
        "+-----------+--------------------+",
        "| Jane      |                    |",
        "+-----------+--------------------+"
      ]

      expected = [
        ["| name      | dob                |"],
        ["| Malcolm   | September 20, 2468 |", "| Reynolds  |                    |"],
        ["| Jane      |                    |"]
      ]

      assert Tabular.cell_line_groups(data_as_lines, has_separators?: true) == expected
    end

    test "without separator lines, extracts lines with cell contents and groups them" do
      data_as_lines = [
        "+-----------+--------------------+",
        "| name      | dob                |",
        "+-----------+--------------------+",
        "| Malcolm   | September 20, 2468 |",
        "| Jane      |                    |",
        "+-----------+--------------------+"
      ]

      expected = [
        ["| name      | dob                |"],
        ["| Malcolm   | September 20, 2468 |"],
        ["| Jane      |                    |"]
      ]

      assert Tabular.cell_line_groups(data_as_lines, has_separators?: false) == expected
    end
  end

  describe "folded_cell_contents" do
    test "returns folded strings" do
      data = [
        [["name", "dob"]],
        [["Malcolm", "September 20, 2468"], ["Reynolds", ""]],
        [["Jane", ""]]
      ]

      expected = [
        ["name", "dob"],
        ["Malcolm Reynolds", "September 20, 2468"],
        ["Jane", ""]
      ]

      assert Tabular.folded_cell_contents(data) == expected
    end
  end

  describe "lines" do
    test "splits and trims lines" do
      data_as_table = """
      line 1
      line 2
      line 3
      """

      expected = [
        "line 1",
        "line 2",
        "line 3"
      ]

      assert Tabular.lines(data_as_table) == expected
    end
  end

  describe "to_list_of_lists" do
    test "returns values as a list of lists" do
      data_as_table = """
            +-----------+--------------------+--------------+
            | name      | dob                | predictable? |
            +-----------+--------------------+--------------+
            | Malcolm   | September 20, 2468 | false        |
            | Reynolds  |                    |              |
            +-----------+--------------------+--------------+
            | Zoe       | February 15, 2484  |              |
            | Washburne |                    |              |
            +-----------+--------------------+--------------+
            | Jane      | :unknown           | true         |
            +-----------+--------------------+--------------+
            | Derrial   | nil                | true         |
            | Book      |                    |              |
            +-----------+--------------------+--------------+
      """

      actual = Tabular.to_list_of_lists(data_as_table)

      expected = [
        ["name", "dob", "predictable?"],
        ["Malcolm Reynolds", "September 20, 2468", false],
        ["Zoe Washburne", "February 15, 2484", ""],
        ["Jane", :unknown, true],
        ["Derrial Book", nil, true]
      ]

      assert actual == expected
    end

    test "returns values as a list of lists from tables without separator lines" do
      data_as_table = """
            +-----------+--------------------+--------------+
            | name      | dob                | predictable? |
            +-----------+--------------------+--------------+
            | Malcolm   | September 20, 2468 | false        |
            | Zoe       | February 15, 2484  |              |
            | Jane      | :unknown           | true         |
            | Derrial   | nil                | true         |
            +-----------+--------------------+--------------+
      """

      actual = Tabular.to_list_of_lists(data_as_table)

      expected = [
        ["name", "dob", "predictable?"],
        ["Malcolm", "September 20, 2468", false],
        ["Zoe", "February 15, 2484", ""],
        ["Jane", :unknown, true],
        ["Derrial", nil, true]
      ]

      assert actual == expected
    end

    test "with header: false, returns values as a list of lists without the header" do
      data_as_table = """
            +-----------+--------------------+--------------+
            | name      | dob                | predictable? |
            +-----------+--------------------+--------------+
            | Malcolm   | September 20, 2468 | false        |
            | Reynolds  |                    |              |
            +-----------+--------------------+--------------+
            | Zoe       | February 15, 2484  |              |
            | Washburne |                    |              |
            +-----------+--------------------+--------------+
            | Jane      | :unknown           | true         |
            +-----------+--------------------+--------------+
            | Derrial   | nil                | true         |
            | Book      |                    |              |
            +-----------+--------------------+--------------+
      """

      actual = Tabular.to_list_of_lists(data_as_table, header: false)

      expected = [
        ["Malcolm Reynolds", "September 20, 2468", false],
        ["Zoe Washburne", "February 15, 2484", ""],
        ["Jane", :unknown, true],
        ["Derrial Book", nil, true]
      ]

      assert actual == expected
    end
  end

  describe "to_list_of_maps" do
    test "returns values as a list of maps" do
      data_as_table = """
            +-----------+--------------------+--------------+
            | name      | dob                | predictable? |
            +-----------+--------------------+--------------+
            | Malcolm   | September 20, 2468 | false        |
            | Reynolds  |                    |              |
            +-----------+--------------------+--------------+
            | Zoe       | February 15, 2484  |              |
            | Washburne |                    |              |
            +-----------+--------------------+--------------+
            | Jane      | :unknown           | true         |
            +-----------+--------------------+--------------+
            | Derrial   | nil                | true         |
            | Book      |                    |              |
            +-----------+--------------------+--------------+
      """

      actual = Tabular.to_list_of_maps(data_as_table)

      expected = [
        %{name: "Malcolm Reynolds", dob: "September 20, 2468", predictable?: false},
        %{name: "Zoe Washburne", dob: "February 15, 2484", predictable?: ""},
        %{name: "Jane", dob: :unknown, predictable?: true},
        %{name: "Derrial Book", dob: nil, predictable?: true}
      ]

      assert actual == expected
    end
  end

  describe "trimmed_and_grouped_cell_contents" do
    test "returns cleaned strings" do
      data_as_line_groups = [
        ["| name      | dob                |"],
        ["| Malcolm   | September 20, 2468 |", "| Reynolds  |                    |"],
        ["| Jane      |                    |"]
      ]

      expected = [
        [["name", "dob"]],
        [["Malcolm", "September 20, 2468"], ["Reynolds", ""]],
        [["Jane", ""]]
      ]

      assert Tabular.trimmed_and_grouped_cell_contents(data_as_line_groups) == expected
    end
  end
end


test/test_support_test.exs
defmodule Tabular.TestSupportTest do
  use ExUnit.Case, async: true

  doctest Tabular.TestSupport

  describe "assert_equal" do
    test "passes when tables are equal" do
      result_table = [
        ["name", "dob"],
        ["Malcolm", "September 20, 2468"],
        ["Zoe", "February 15, 2484"]
      ]

      assert Tabular.TestSupport.assert_equal(result_table)
    end

    test "flunks and prints table highlighting differences when tables are different" do
      expected_message = ~r|Malcolm <=> Mike|

      results_table = [
        ["name", "dob"],
        [{"Malcolm", "Mike"}, "September 20, 2468"],
        ["Zoe", "February 15, 2484"]
      ]

      assert_raise ExUnit.AssertionError, expected_message, fn ->
        Tabular.TestSupport.assert_equal(results_table)
      end
    end
  end

  describe "equal?" do
    test "when there is a non-matching value in any cell, returns false" do
      results_table = [
        ["name", "dob"],
        [{"Malcolm", "Mike"}, "September 20, 2468"],
        ["Zoe", "February 15, 2484"]
      ]

      refute Tabular.TestSupport.equal?(results_table)
    end

    test "when all cells match, returns true" do
      results_table = [
        ["name", "dob"],
        ["Mike", "September 20, 2468"],
        ["Zoe", "February 15, 2484"]
      ]

      assert Tabular.TestSupport.equal?(results_table)
    end
  end

  describe "compare" do
    test "creates a matrix indicating non-matching cells with `false`" do
      table1 = """
            +---------+--------------------+
            | name    | dob                |
            +---------+--------------------+
            | Malcolm | September 20, 2468 |
            +---------+--------------------+
            | Zoe     | February 15, 2484  |
            +---------+--------------------+
      """

      table2 = """
            +---------+--------------------+
            | name    | dob                |
            +---------+--------------------+
            | Mike    | September 20, 2468 |
            +---------+--------------------+
            | Zoe     | February 15, 2484  |
            +---------+--------------------+
      """

      results_table = [
        ["name", "dob"],
        [{"Malcolm", "Mike"}, "September 20, 2468"],
        ["Zoe", "February 15, 2484"]
      ]

      assert Tabular.TestSupport.compare(table1, table2) == results_table
    end

    test "uses comparator functions to compare cells" do
      table1 = """
            +---------+-------+
            | name    | count |
            +---------+-------+
            | Malcolm | 10    |
            +---------+-------+
            | Zoe     | 5     |
            +---------+-------+
      """

      table2 = """
            +---------+-------+
            | name    | count |
            +---------+-------+
            | Mike    | 11    |
            +---------+-------+
            | Zoe     | 20    |
            +---------+-------+
      """

      results_table = [
        ["name", "count"],
        [{"Malcolm", "Mike"}, "10"],
        ["Zoe", {"5", "20"}]
      ]

      comparators = %{"count" => &(abs(String.to_integer(&1) - String.to_integer(&2)) < 2)}

      assert Tabular.TestSupport.compare(table1, table2, comparators: comparators) ==
               results_table
    end
  end
end


lib/tabular/test_support.ex
defmodule Tabular.TestSupport do
  @moduledoc """
  Functions to simplify testing with ascii tables.
  """

  @doc ~S'''

  Compares two ascii tables, producing a matrix of the individual comparisons.

  An optional map of comparator functions can be used. Comparators are useful when values
  in the tables cannot be compared by simple equality, such as floating point values that
  may be consider equal when the 2 values are close.

  Each cell
  in the matrix is either a single value or a tuple of values.

  - A single value indicates that both tables have the same value at that position. The
    value is the actual value found in both tables at that position or the value from the
    first table when using a comparator function

  - A tuple indicates that the two tables have different values at that position. The
    tuple contains the differing values from the two tables.

  Comparators are functions that take 2 values as arguments and return true when the
  values are to be considered a match and false otherwise.

  The keys in the comparator map indicate the table column where the comparator is to be
  used.

  ## Example

      iex> table1 = """
      ...>   +---------+-------+
      ...>   | name    | count |
      ...>   +---------+-------+
      ...>   | Malcolm | 10    |
      ...>   +---------+-------+
      ...>   | Zoe     | 5     |
      ...>   +---------+-------+
      ...> """
      ...>
      ...> table2 = """
      ...>   +---------+-------+
      ...>   | name    | count |
      ...>   +---------+-------+
      ...>   | Mike    | 11    |
      ...>   +---------+-------+
      ...>   | Zoe     | 20    |
      ...>   +---------+-------+
      ...> """
      ...>
      ...> comparators = %{"count" => &(abs(String.to_integer(&1) - String.to_integer(&2)) < 2)}
      ...>
      ...> Tabular.TestSupport.compare(table1, table2, comparators: comparators)
      [
        ["name", "count"],
        [{"Malcolm", "Mike"}, "10"],
        ["Zoe", {"5", "20"}]
      ]
  '''

  def compare(actual_table, expected_table, opts \\ [comparators: %{}]) do
    actual_table_data = actual_table |> Tabular.to_list_of_lists()
    expected_table_data = expected_table |> Tabular.to_list_of_lists()

    [actual_headers | actual_table_rows] = actual_table_data
    [expected_header_row | expected_table_rows] = expected_table_data

    [actual_headers] ++
      (Enum.zip(actual_table_rows, expected_table_rows)
       |> Enum.map(fn {actual_row, expected_row} ->
         Enum.zip([expected_header_row, actual_row, expected_row])
         |> Enum.map(fn {header, actual_value, expected_value} ->
           match? =
             case opts[:comparators][header] do
               nil -> actual_value == expected_value
               comparator -> comparator.(actual_value, expected_value)
             end

           case match? do
             true -> actual_value
             false -> {actual_value, expected_value}
           end
         end)
       end))
  end

  @doc ~S'''

  Examines the matrix produced by `compare`. Returns false
  if any cell is false. Otherwise returns true.

  ## Examples

      iex> results_table =
      ...> [
      ...>   ["name", "count"],
      ...>   [{"Malcolm", "Mike"}, "10"],
      ...>   ["Zoe", {"5", "20"}]
      ...> ]
      ...>
      ...> Tabular.TestSupport.equal?(results_table)
      false
  '''

  def equal?(compare_result) do
    Enum.all?(compare_result, fn row ->
      Enum.all?(row, fn
        {_, _} -> false
        _ -> true
      end)
    end)
  end

  @doc ~S'''

  Asserts that the contents of 2 tables are equal.
  If the tables are not equal, the failure message
  includes the table highlighting the differences
  in the cells that disagree.

  ## Examples

      iex> try do
      ...>   [
      ...>     ["name", "count"],
      ...>     [{"Malcolm", "Mike"}, "10"],
      ...>     ["Zoe", {"5", "20"}]
      ...>   ]
      ...>   |> Tabular.TestSupport.assert_equal()
      ...> rescue
      ...>   e in ExUnit.AssertionError ->
      ...>     e.message
      ...> end
      """
      +--------------------------+------------------+
      | name                     | count            |
      +--------------------------+------------------+
      | >>> Malcolm <=> Mike <<< | 10               |
      | Zoe                      | >>> 5 <=> 20 <<< |
      +--------------------------+------------------+
      """
  '''

  def assert_equal(results_table) do
    if equal?(results_table) do
      true
    else
      generate_error_message(results_table)
      |> ExUnit.Assertions.flunk()
    end
  end

  @doc false
  def generate_error_message([header | body]) do
    format_cells(body) |> TableRex.quick_render!(header)
  end

  @doc false
  def format_cells(results_table) do
    Enum.map(results_table, fn row ->
      Enum.map(row, fn
        {left, right} -> ">>> #{left} <=> #{right} <<<"
        cell -> cell
      end)
    end)
  end
end


mix.exs
defmodule Tabular.MixProject do
  use Mix.Project

  def project do
    [
      app: :tabular,
      description: description(),
      version: "1.2.1",
      elixir: "~> 1.8",
      deps: deps(),
      name: "Tabular",
      source_url: "https://github.com/kellyfelkins/tabular",
      package: package(),
      docs: [
        main: "Tabular",
        extras: ["README.md", "CHANGELOG.md"]
      ]
    ]
  end

  defp deps do
    [
      {:ex_doc, "~> 0.19", only: :dev, runtime: false},
      {:table_rex, "~> 2.0.0"}
    ]
  end

  defp description do
    "Tabular converts an ascii table string into either a list of lists or a list of maps"
  end

  defp package do
    [
      licenses: ["BlueOak-1.0.0"],
      links: %{
        "GitHub" => "https://github.com/kellyfelkins/tabular"
      }
    ]
  end

  def application do
    [applications: [:table_rex]]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# third-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :tabular, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:tabular, :key)
#
# You can also configure a third-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env()}.exs"


lib/tabular.ex
defmodule Tabular do
  @moduledoc """
  Tabular converts an ascii table string into either a list of lists or a list of maps.
  """

  @row_splitter_re ~r/^\s*[|+]--[-+|]+$/

  @doc ~S'''

  Converts an ascii table to a list of maps.

  ## Examples

      iex> ascii_table = """
      ...> |---------------+--------------------|
      ...> | name          | dob                |
      ...> |---------------+--------------------|
      ...> | Malcolm       | September 20, 2468 |
      ...> | Reynolds      |                    |
      ...> |---------------+--------------------|
      ...> | Zoe Washburne | February 15, 2484  |
      ...> |---------------+--------------------|
      ...> """
      ...> Tabular.to_list_of_maps(ascii_table)
      [
        %{dob: "September 20, 2468", name: "Malcolm Reynolds"},
        %{dob: "February 15, 2484", name: "Zoe Washburne"}
      ]

      Without separators:

      iex> ascii_table = """
      ...> |------------------+--------------------|
      ...> | name             | dob                |
      ...> |------------------+--------------------|
      ...> | Malcolm Reynolds | September 20, 2468 |
      ...> | Zoe Washburne    | February 15, 2484  |
      ...> |------------------+--------------------|
      ...> """
      ...> Tabular.to_list_of_maps(ascii_table)
      [
        %{dob: "September 20, 2468", name: "Malcolm Reynolds"},
        %{dob: "February 15, 2484", name: "Zoe Washburne"}
      ]

  '''
  def to_list_of_maps(ascii_table) do
    [headers | rows] = to_list_of_lists(ascii_table)
    headers_as_atoms = Enum.map(headers, &String.to_atom(&1))

    rows
    |> Enum.map(fn row ->
      Enum.zip(headers_as_atoms, row)
      |> Map.new()
    end)
  end

  @doc deprecated: "Use to_list_of_lists(table, header: false) instead"
  @doc ~S'''

  Converts an ascii table to a list of lists, omitting the header row.

  ## Examples

      iex> ascii_table = """
      ...> |---------------+--------------------|
      ...> | name          | dob                |
      ...> |---------------+--------------------|
      ...> | Malcolm       | September 20, 2468 |
      ...> | Reynolds      |                    |
      ...> |---------------+--------------------|
      ...> | Zoe Washburne | February 15, 2484  |
      ...> |---------------+--------------------|
      ...> """
      ...> Tabular.to_list_of_lists_no_header(ascii_table)
      [
        ["Malcolm Reynolds", "September 20, 2468"],
        ["Zoe Washburne", "February 15, 2484"]
      ]

      Without separators:

      iex> ascii_table = """
      ...> |------------------+--------------------|
      ...> | name             | dob                |
      ...> |------------------+--------------------|
      ...> | Malcolm Reynolds | September 20, 2468 |
      ...> | Zoe Washburne    | February 15, 2484  |
      ...> |------------------+--------------------|
      ...> """
      ...> Tabular.to_list_of_lists_no_header(ascii_table)
      [
        ["Malcolm Reynolds", "September 20, 2468"],
        ["Zoe Washburne", "February 15, 2484"]
      ]

  '''
  def to_list_of_lists_no_header(ascii_table) do
    to_list_of_lists(ascii_table, header: false)
  end

  @doc ~S'''

  Converts an ascii table to a list of lists.

  ## Examples

      iex> ascii_table = """
      ...> |---------------+--------------------|
      ...> | name          | dob                |
      ...> |---------------+--------------------|
      ...> | Malcolm       | September 20, 2468 |
      ...> | Reynolds      |                    |
      ...> |---------------+--------------------|
      ...> | Zoe Washburne | February 15, 2484  |
      ...> |---------------+--------------------|
      ...> """
      ...> Tabular.to_list_of_lists(ascii_table)
      [
        ["name", "dob"],
        ["Malcolm Reynolds", "September 20, 2468"],
        ["Zoe Washburne", "February 15, 2484"]
      ]
      ...> Tabular.to_list_of_lists(ascii_table, header: false)
      [
        ["Malcolm Reynolds", "September 20, 2468"],
        ["Zoe Washburne", "February 15, 2484"]
      ]

      Without separators:

      iex> ascii_table = """
      ...> |------------------+--------------------|
      ...> | name             | dob                |
      ...> |------------------+--------------------|
      ...> | Malcolm Reynolds | September 20, 2468 |
      ...> | Zoe Washburne    | February 15, 2484  |
      ...> |------------------+--------------------|
      ...> """
      ...> Tabular.to_list_of_lists(ascii_table)
      [
        ["name", "dob"],
        ["Malcolm Reynolds", "September 20, 2468"],
        ["Zoe Washburne", "February 15, 2484"]
      ]
      ...> Tabular.to_list_of_lists(ascii_table, header: false)
      [
        ["Malcolm Reynolds", "September 20, 2468"],
        ["Zoe Washburne", "February 15, 2484"]
      ]
  '''
  def to_list_of_lists(ascii_table, opts \\ [header: true]) do
    rows =
      ascii_table
      |> lines()
      |> cell_line_groups(has_separators?: has_separators?(ascii_table))
      |> trimmed_and_grouped_cell_contents()
      |> folded_cell_contents()
      |> specials()

    if opts[:header], do: rows, else: Enum.drop(rows, 1)
  end

  @doc false
  def lines(ascii_table) do
    ascii_table
    |> String.trim()
    |> String.split("\n")
  end

  @doc false
  def cell_line_groups(lines, has_separators?: true) do
    lines
    |> Enum.chunk_by(fn line ->
      Regex.match?(@row_splitter_re, line)
    end)
    |> Enum.filter(fn group ->
      !(length(group) == 1 &&
          Regex.match?(@row_splitter_re, hd(group)))
    end)
  end

  @doc false
  def cell_line_groups(lines, has_separators?: false) do
    lines
    |> Enum.reject(&(&1 =~ @row_splitter_re))
    |> Enum.map(&[&1])
  end

  @doc false
  def trimmed_and_grouped_cell_contents(line_groups) do
    line_groups
    |> Enum.map(fn line_group ->
      line_group
      |> Enum.map(fn line ->
        [_all, trimmed_line] = Regex.run(~r/^\s*\|\s*(.*)\s*\|\s*$/, line)

        trimmed_line
        |> String.split("|")
        |> Enum.map(fn cell_content ->
          String.trim(cell_content)
        end)
      end)
    end)
  end

  @doc false
  def folded_cell_contents(grouped_cell_contents) do
    grouped_cell_contents
    |> Enum.map(fn row_contents ->
      row_contents
      |> fold_contents_if_necessary
    end)
  end

  @doc false
  def fold_contents_if_necessary(row_contents) when length(row_contents) > 1 do
    Enum.zip(row_contents)
    |> Enum.map(fn foldables ->
      foldables
      |> Tuple.to_list()
      |> Enum.filter(fn text -> text != "" end)
      |> Enum.join(" ")
    end)
  end

  @doc false
  def fold_contents_if_necessary(row_contents) do
    hd(row_contents)
  end

  @doc false
  def specials(rows_of_columns) do
    rows_of_columns
    |> Enum.map(fn row ->
      row
      |> Enum.map(fn text ->
        special(text)
      end)
    end)
  end

  @doc false
  def special("true"), do: true
  def special("false"), do: false
  def special("nil"), do: nil
  def special(""), do: ""
  def special(":" <> rest), do: String.to_atom(rest)
  def special(not_special), do: not_special

  @doc false
  def has_separators?(ascii_table) do
    [_, _, _ | lines] = lines(ascii_table)
    {_, lines} = List.pop_at(lines, -1)

    Enum.any?(lines, &(&1 =~ @row_splitter_re))
  end
end
